File: llm-tools/antidote/tools/llm-cli.py
#!/usr/bin/env python3
"""
LLM Antidote CLI Tool

Author: Jonathan Jewell
Purpose: Command-line interface for managing reset artifacts and diagnostics

Usage:
  llm-cli.py reset [universal|coding|conversation]
  llm-cli.py preserve
  llm-cli.py verify
  llm-cli.py diagnose [memory|continuity|association]
  llm-cli.py list
  llm-cli.py info <artifact>
"""

import os
import sys
import json
from pathlib import Path
from typing import Optional, List, Dict


class ArtifactManager:
    """Manages LLM reset and preservation artifacts."""

    def __init__(self, repo_root: Optional[Path] = None):
        """
        Initialize artifact manager.

        Args:
            repo_root: Root directory of llm-antidote repository
        """
        if repo_root is None:
            # Try to find repo root
            current = Path.cwd()
            while current != current.parent:
                if (current / "llm-reset").exists():
                    repo_root = current
                    break
                current = current.parent
            else:
                repo_root = Path.cwd()

        self.repo_root = repo_root
        self.artifacts_dir = repo_root / "artifacts"
        self.tools_dir = repo_root / "tools"

    def list_artifacts(self) -> List[Dict[str, str]]:
        """
        List all available artifacts.

        Returns:
            List of artifact metadata
        """
        artifacts = []

        # Check root directory
        if (self.repo_root / "llm-reset").exists():
            artifacts.append({
                "name": "universal",
                "file": "llm-reset",
                "format": "scheme",
                "description": "Universal semantic reset (original)"
            })

        # Check artifacts directory
        if self.artifacts_dir.exists():
            artifact_map = {
                "llm-reset.py": ("universal-py", "python", "Universal reset (Python)"),
                "llm-reset.js": ("universal-js", "javascript", "Universal reset (JavaScript)"),
                "llm-reset.json": ("universal-json", "json", "Universal reset (JSON)"),
                "llm-reset-coding.scm": ("coding", "scheme", "Coding context reset"),
                "llm-reset-conversation.scm": ("conversation", "scheme", "Conversation thread reset"),
                "llm-preserve.scm": ("preserve", "scheme", "Context preservation"),
            }

            for filename, (name, fmt, desc) in artifact_map.items():
                if (self.artifacts_dir / filename).exists():
                    artifacts.append({
                        "name": name,
                        "file": f"artifacts/{filename}",
                        "format": fmt,
                        "description": desc
                    })

        return artifacts

    def get_artifact(self, name: str) -> Optional[str]:
        """
        Get artifact content by name.

        Args:
            name: Artifact name (e.g., 'universal', 'coding', 'preserve')

        Returns:
            Artifact content or None if not found
        """
        artifacts = self.list_artifacts()
        for artifact in artifacts:
            if artifact["name"] == name:
                file_path = self.repo_root / artifact["file"]
                if file_path.exists():
                    return file_path.read_text()
        return None

    def get_diagnostic(self, diagnostic_type: str) -> Optional[str]:
        """
        Get diagnostic prompt.

        Args:
            diagnostic_type: Type of diagnostic (memory|continuity|association|verify)

        Returns:
            Diagnostic prompt or None
        """
        from tools.llm_diagnostic import ContextProbe

        probe = ContextProbe()

        if diagnostic_type == "memory":
            return probe.generate_memory_probe()
        elif diagnostic_type == "continuity":
            return probe.generate_continuity_probe()
        elif diagnostic_type == "verify":
            return probe.generate_reset_verification()
        elif diagnostic_type == "association":
            return probe.generate_association_probe("context")
        else:
            return None


def print_artifact(content: str):
    """Print artifact content with formatting."""
    print("\n" + "=" * 70)
    print(content)
    print("=" * 70 + "\n")
    print("📋 Copy the above artifact and paste it into your LLM chat.")
    print()


def print_artifact_list(artifacts: List[Dict[str, str]]):
    """Print formatted artifact list."""
    print("\n📚 Available Artifacts:\n")
    for i, artifact in enumerate(artifacts, 1):
        print(f"{i}. {artifact['name']:<20} ({artifact['format']:<10}) - {artifact['description']}")
    print()


def print_usage():
    """Print usage information."""
    print("""
🔧 LLM Antidote CLI Tool

Usage:
  llm-cli.py reset [TYPE]       Display reset artifact
  llm-cli.py preserve           Display context preservation artifact
  llm-cli.py verify             Display reset verification tool
  llm-cli.py diagnose [TYPE]    Display diagnostic prompt
  llm-cli.py list               List all artifacts
  llm-cli.py info <artifact>    Show artifact information

Reset Types:
  universal       Universal semantic reset (default)
  universal-py    Universal reset (Python)
  universal-js    Universal reset (JavaScript)
  universal-json  Universal reset (JSON)
  coding          Coding context reset
  conversation    Conversation thread reset

Diagnostic Types:
  memory          Test what LLM remembers
  continuity      Test conversation continuity
  verify          Verify reset effectiveness
  association     Test word associations

Examples:
  llm-cli.py reset                    # Show universal reset
  llm-cli.py reset coding             # Show coding context reset
  llm-cli.py preserve                 # Show preservation artifact
  llm-cli.py diagnose memory          # Show memory diagnostic
  llm-cli.py list                     # List all artifacts

Workflow:
  1. Use 'reset' to clear LLM context
  2. Use 'verify' to check if reset worked
  3. Use 'preserve' to mark important context
  4. Use 'diagnose' to probe context state
""")


def main():
    """CLI entry point."""
    manager = ArtifactManager()

    if len(sys.argv) < 2:
        print_usage()
        return

    command = sys.argv[1]

    if command == "list":
        print_artifact_list(manager.list_artifacts())

    elif command == "reset":
        artifact_type = sys.argv[2] if len(sys.argv) > 2 else "universal"
        content = manager.get_artifact(artifact_type)
        if content:
            print_artifact(content)
        else:
            print(f"❌ Artifact '{artifact_type}' not found.")
            print("\n💡 Use 'llm-cli.py list' to see available artifacts.")

    elif command == "preserve":
        content = manager.get_artifact("preserve")
        if content:
            print_artifact(content)
        else:
            print("❌ Preservation artifact not found.")

    elif command == "verify":
        verify_path = manager.tools_dir / "llm-verify.scm"
        if verify_path.exists():
            print_artifact(verify_path.read_text())
        else:
            print("❌ Verification tool not found.")

    elif command == "diagnose":
        diagnostic_type = sys.argv[2] if len(sys.argv) > 2 else "memory"
        content = manager.get_diagnostic(diagnostic_type)
        if content:
            print_artifact(content)
        else:
            print(f"❌ Diagnostic '{diagnostic_type}' not found.")
            print("\n💡 Available: memory, continuity, verify, association")

    elif command == "info":
        if len(sys.argv) < 3:
            print("❌ Please specify an artifact name.")
            print("💡 Use 'llm-cli.py list' to see available artifacts.")
        else:
            artifact_name = sys.argv[2]
            artifacts = manager.list_artifacts()
            for artifact in artifacts:
                if artifact["name"] == artifact_name:
                    print(f"\n📄 Artifact: {artifact['name']}")
                    print(f"   File: {artifact['file']}")
                    print(f"   Format: {artifact['format']}")
                    print(f"   Description: {artifact['description']}\n")
                    return
            print(f"❌ Artifact '{artifact_name}' not found.")

    else:
        print(f"❌ Unknown command: {command}")
        print_usage()


if __name__ == "__main__":
    main()

File: llm-tools/antidote/tools/llm_cli.jl
#!/usr/bin/env julia
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath)

"""
    LLM Antidote CLI Tool

    Author: Jonathan D.A. Jewell
    Purpose: Command-line interface for managing reset artifacts and diagnostics

    Usage:
      llm_cli.jl reset [universal|coding|conversation]
      llm_cli.jl preserve
      llm_cli.jl verify
      llm_cli.jl diagnose [memory|continuity|association]
      llm_cli.jl list
      llm_cli.jl info <artifact>
"""

# ---------------------------------------------------------------------------
# Artifact Manager
# ---------------------------------------------------------------------------

"""
    ArtifactManager

Manages LLM reset and preservation artifacts by scanning the repository
directory structure for known artifact files and providing access to their
contents.
"""
struct ArtifactManager
    "Root directory of the llm-antidote repository"
    repo_root::String
    "Path to the artifacts/ subdirectory"
    artifacts_dir::String
    "Path to the tools/ subdirectory"
    tools_dir::String
end

"""
    ArtifactManager(; repo_root=nothing)

Construct an ArtifactManager, optionally specifying the repo root. If not
provided, walks up from cwd looking for the `llm-reset` file to find the
repository root.
"""
function ArtifactManager(; repo_root::Union{String,Nothing}=nothing)
    if isnothing(repo_root)
        # Walk up from cwd to find repo root containing llm-reset
        current = pwd()
        while true
            if isfile(joinpath(current, "llm-reset"))
                repo_root = current
                break
            end
            parent = dirname(current)
            if parent == current
                # Reached filesystem root without finding it
                repo_root = pwd()
                break
            end
            current = parent
        end
    end

    return ArtifactManager(
        repo_root,
        joinpath(repo_root, "artifacts"),
        joinpath(repo_root, "tools")
    )
end

"""
    list_artifacts(manager)

List all available artifacts with their metadata (name, file path, format,
and description).

# Returns
- `Vector{Dict{String,String}}`: list of artifact metadata dictionaries
"""
function list_artifacts(manager::ArtifactManager)::Vector{Dict{String,String}}
    artifacts = Dict{String,String}[]

    # Check root directory for the universal reset artifact
    if isfile(joinpath(manager.repo_root, "llm-reset"))
        push!(artifacts, Dict(
            "name" => "universal",
            "file" => "llm-reset",
            "format" => "scheme",
            "description" => "Universal semantic reset (original)"
        ))
    end

    # Check artifacts directory for variant artifacts
    if isdir(manager.artifacts_dir)
        artifact_map = Dict{String,Tuple{String,String,String}}(
            "llm-reset.py" => ("universal-py", "python", "Universal reset (Python)"),
            "llm-reset.js" => ("universal-js", "javascript", "Universal reset (JavaScript)"),
            "llm-reset.json" => ("universal-json", "json", "Universal reset (JSON)"),
            "llm-reset-coding.scm" => ("coding", "scheme", "Coding context reset"),
            "llm-reset-conversation.scm" => ("conversation", "scheme", "Conversation thread reset"),
            "llm-preserve.scm" => ("preserve", "scheme", "Context preservation")
        )

        for (filename, (artifact_name, format, description)) in artifact_map
            if isfile(joinpath(manager.artifacts_dir, filename))
                push!(artifacts, Dict(
                    "name" => artifact_name,
                    "file" => "artifacts/$filename",
                    "format" => format,
                    "description" => description
                ))
            end
        end
    end

    return artifacts
end

"""
    get_artifact(manager, name)

Retrieve the content of a named artifact.

# Arguments
- `manager::ArtifactManager`: the artifact manager
- `name::String`: artifact name (e.g., "universal", "coding", "preserve")

# Returns
- `Union{String,Nothing}`: artifact content, or nothing if not found
"""
function get_artifact(manager::ArtifactManager, name::String)::Union{String,Nothing}
    artifacts = list_artifacts(manager)
    for artifact in artifacts
        if artifact["name"] == name
            file_path = joinpath(manager.repo_root, artifact["file"])
            if isfile(file_path)
                return read(file_path, String)
            end
        end
    end
    return nothing
end

"""
    get_diagnostic(manager, diagnostic_type)

Generate a diagnostic prompt of the given type by delegating to the
ContextProbe from llm_diagnostic.jl.

# Arguments
- `manager::ArtifactManager`: the artifact manager
- `diagnostic_type::String`: one of "memory", "continuity", "verify", "association"

# Returns
- `Union{String,Nothing}`: the diagnostic prompt, or nothing if type is unknown
"""
function get_diagnostic(manager::ArtifactManager, diagnostic_type::String)::Union{String,Nothing}
    # Import diagnostic functions from the sibling module
    diagnostic_path = joinpath(manager.tools_dir, "llm_diagnostic.jl")
    if isfile(diagnostic_path)
        include(diagnostic_path)
    end

    probe = ContextProbe()

    if diagnostic_type == "memory"
        return generate_memory_probe(probe)
    elseif diagnostic_type == "continuity"
        return generate_continuity_probe(probe)
    elseif diagnostic_type == "verify"
        return generate_reset_verification(probe)
    elseif diagnostic_type == "association"
        return generate_association_probe(probe, "context")
    else
        return nothing
    end
end

# ---------------------------------------------------------------------------
# Output Formatting
# ---------------------------------------------------------------------------

"""
    print_artifact(content)

Print artifact content wrapped in separator lines with a copy instruction.
"""
function print_artifact(content::String)
    println()
    println(repeat("=", 70))
    println(content)
    println(repeat("=", 70))
    println()
    println("Copy the above artifact and paste it into your LLM chat.")
    println()
end

"""
    print_artifact_list(artifacts)

Print a formatted table of available artifacts.
"""
function print_artifact_list(artifacts::Vector{Dict{String,String}})
    println("\nAvailable Artifacts:\n")
    for (index, artifact) in enumerate(artifacts)
        padded_name = rpad(artifact["name"], 20)
        padded_format = rpad("($(artifact["format"]))", 12)
        println("$index. $padded_name $padded_format - $(artifact["description"])")
    end
    println()
end

# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

"""
    print_usage()

Print CLI usage information.
"""
function print_usage()
    println("""

LLM Antidote CLI Tool

Usage:
  llm_cli.jl reset [TYPE]       Display reset artifact
  llm_cli.jl preserve           Display context preservation artifact
  llm_cli.jl verify             Display reset verification tool
  llm_cli.jl diagnose [TYPE]    Display diagnostic prompt
  llm_cli.jl list               List all artifacts
  llm_cli.jl info <artifact>    Show artifact information

Reset Types:
  universal       Universal semantic reset (default)
  universal-py    Universal reset (Python)
  universal-js    Universal reset (JavaScript)
  universal-json  Universal reset (JSON)
  coding          Coding context reset
  conversation    Conversation thread reset

Diagnostic Types:
  memory          Test what LLM remembers
  continuity      Test conversation continuity
  verify          Verify reset effectiveness
  association     Test word associations

Examples:
  llm_cli.jl reset                    # Show universal reset
  llm_cli.jl reset coding             # Show coding context reset
  llm_cli.jl preserve                 # Show preservation artifact
  llm_cli.jl diagnose memory          # Show memory diagnostic
  llm_cli.jl list                     # List all artifacts

Workflow:
  1. Use 'reset' to clear LLM context
  2. Use 'verify' to check if reset worked
  3. Use 'preserve' to mark important context
  4. Use 'diagnose' to probe context state
""")
end

"""
    main()

CLI entry point. Supports subcommands: list, reset [type], preserve, verify,
diagnose [type], info <name>.
"""
function main()
    manager = ArtifactManager()

    if length(ARGS) < 1
        print_usage()
        return
    end

    command = ARGS[1]

    if command == "list"
        print_artifact_list(list_artifacts(manager))

    elseif command == "reset"
        artifact_type = length(ARGS) > 1 ? ARGS[2] : "universal"
        content = get_artifact(manager, artifact_type)
        if !isnothing(content)
            print_artifact(content)
        else
            println("Artifact '$artifact_type' not found.")
            println("\nUse 'llm_cli.jl list' to see available artifacts.")
        end

    elseif command == "preserve"
        content = get_artifact(manager, "preserve")
        if !isnothing(content)
            print_artifact(content)
        else
            println("Preservation artifact not found.")
        end

    elseif command == "verify"
        verify_path = joinpath(manager.tools_dir, "llm-verify.scm")
        if isfile(verify_path)
            print_artifact(read(verify_path, String))
        else
            println("Verification tool not found.")
        end

    elseif command == "diagnose"
        diagnostic_type = length(ARGS) > 1 ? ARGS[2] : "memory"
        content = get_diagnostic(manager, diagnostic_type)
        if !isnothing(content)
            print_artifact(content)
        else
            println("Diagnostic '$diagnostic_type' not found.")
            println("\nAvailable: memory, continuity, verify, association")
        end

    elseif command == "info"
        if length(ARGS) < 2
            println("Please specify an artifact name.")
            println("Use 'llm_cli.jl list' to see available artifacts.")
        else
            artifact_name = ARGS[2]
            artifacts = list_artifacts(manager)
            found = false
            for artifact in artifacts
                if artifact["name"] == artifact_name
                    println("\nArtifact: $(artifact["name"])")
                    println("   File: $(artifact["file"])")
                    println("   Format: $(artifact["format"])")
                    println("   Description: $(artifact["description"])\n")
                    found = true
                    break
                end
            end
            if !found
                println("Artifact '$artifact_name' not found.")
            end
        end

    else
        println("Unknown command: $command")
        print_usage()
    end
end

# Run main when executed as a script
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
